### PR TITLE
New checks

### PR DESCRIPTION
**Pull request by huguesBouvier**
1. Added 1 check: When in nested configuration pool.ntp.org:123 might not be available and IoTedge will connect to a parent IoTedge and not to IoT Hub.
The time is the checked directly against the parent IoT edge. The check is performed by making an HTTPS and the reading the "date" header from the response.

2. Improved parent hostname check in container. Adding extrahosts from agent to diagnostic image when performing parent hostname resolution

3. Changed API-config template so API proxy can route all types of blob request (instead of just append blob type).

4. Fixed a bug in API proxy script to build the image

Test:
1. Checked without parent hostname and checked all test pass (no regression check)
2. Checked local time check by displaying the offset in milliseconds first.
3. Check add extrahost by hard coding "dummy" has the parent hostname in the code adding extahost=dummy:"parent ip adress" and checked it worked. Then changed the IP with a dummy IP and checked it failed.

### CODE REVIEW

File: edgelet/docker-rs/src/models/host_config.rs [hunk 1]
```diff
@@ -184,8 +184,8 @@ pub struct HostConfig {
     // #[serde(rename = "DnsSearch", skip_serializing_if = "Option::is_none")]
     // dns_search: Option<Vec<String>>,
     // /// A list of hostnames/IP mappings to add to the container's `/etc/hosts` file. Specified in the form `[\"hostname:IP\"]`.
-    // #[serde(rename = "ExtraHosts", skip_serializing_if = "Option::is_none")]
-    // extra_hosts: Option<Vec<String>>,
+    #[serde(rename = "ExtraHosts", skip_serializing_if = "Option::is_none")]
```
Comment (varunpuranik) [inline edgelet/docker-rs/src/models/host_config.rs:187]: What scenario does this enable? I believe specifying extrahosts in createOptions was already working?

File: edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs [hunk 2]
```diff
@@ -49,27 +55,33 @@ impl ContainerResolveParentHostname {
...
+            .agent()
+            .config()
+            .create_options()
+            .host_config()
+            .and_then(docker::models::HostConfig::extra_hosts)
+            .iter()
+            .for_each(|extra_hosts| {
+                extra_hosts
+                    .iter()
+                    .for_each(|host| args.push(format!("--add-host={}", host)))
```
Comment (varunpuranik) [inline edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs:70]: I suppose this is why you need to uncomment extraHosts above?

File: edgelet/iotedge/src/check/checks/pull_agent_from_upstream.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+#[derive(Default, serde_derive::Serialize)]
+pub(crate) struct PullAgentFromUpstream {}
+
+impl Checker for PullAgentFromUpstream {
+    fn id(&self) -> &'static str {
+        "Pull-Agent-from-upstream"
+    }
+    fn description(&self) -> &'static str {
+        "config.yaml edge Agent can be pulled from upstream"
```
Comment (varunpuranik) [inline edgelet/iotedge/src/check/checks/pull_agent_from_upstream.rs:13]:
```suggestion
        "EdgeAgent module can be pulled from upstream"
```

**Comment by varunpuranik** [comment]
Nit: can you please add a more descriptive title and description - what checks you are adding, and what is the approach.

File: edgelet/docker-rs/src/models/host_config.rs (hunk 1 above)
Comment (huguesBouvier) [reply edgelet/docker-rs/src/models/host_config.rs:187]: The scenario is the following:
We read the extra host give to the agent
We run the diagnostic module with the "extrahost" parameter and we resolve the hostname.

To be able to read extra host we need to enable this.

File: edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs (hunk 2 above)
Comment (huguesBouvier) [reply edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs:70]: Correct :)

**Comment by huguesBouvier** [comment]
> it: can you please add a more descriptive title and description - what checks you are addi

Yes, the code is completed. I still need to do some stuff:
1. better description
2. more complete testing
3. rebasing

**Review by lfitchett — APPROVED** [review]

**Pull request merged**